src/agent/rustjail/src/reaper.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2020 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use nix::fcntl::OFlag;
+use slog::Logger;
+
+use nix::unistd;
+use std::os::unix::io::RawFd;
+
+use anyhow::Result;
+
+const MAX_EVENTS: usize = 2;
+
+#[derive(Debug, Clone)]
+pub struct Epoller {
+    logger: Logger,
+    epoll_fd: RawFd,
+    // rfd and wfd are a pipe's files two ends, this pipe is
+    // used to sync between the readStdio and the process exits.
+    // once the process exits, it will close one end to notify
+    // the readStdio that the process has exited and it should not
+    // wait on the process's terminal which has been inherited
+    // by it's children and hasn't exited.
+    rfd: RawFd,
+    wfd: RawFd,
+}
+
+impl Epoller {
+    pub fn new(logger: &Logger, fd: RawFd) -> Result<Epoller> {
+        let epoll_fd = epoll::create(true)?;
+        let (rfd, wfd) = unistd::pipe2(OFlag::O_CLOEXEC)?;
+
+        let mut epoller = Self {
+            logger: logger.clone(),
+            epoll_fd,
+            rfd,
+            wfd,
+        };
+
+        epoller.add(rfd)?;
+        epoller.add(fd)?;
+
+        Ok(epoller)
+    }
+
+    pub fn close_wfd(&self) {
+        let _ = unistd::close(self.wfd);
+    }
+
+    // pub fn close_rfd(&self) {
+    //     let _ = unistd::close(self.rfd);
+    // }
+
+    pub fn close(&self) {
```

Comment:
Would it make sense to impl `Drop` trait for this? So there is no way we forget to call this and hence no descriptor leaks?